{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORT THE NECESSARY LIBRARIES\n",
    "import pandas as pd\n",
    "import re # This module provides functions for working with regular expressions allowing search, match and manipulate strings based on patterns\n",
    "import os # This module allows interaction with the operating system, such as working with files, directories, environment variables, and system commands.\n",
    "import sys # This module provides access to system-specific parameters, including command-line arguments (sys.argv), standard input/output, and exiting the program (sys.exit()).\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.optimize as opt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from statsmodels.graphics.tsaplots import plot_acf # This function is used for plotting the Autocorrelation Function (ACF), which is helpful in time series analysis to measure how observations are correlated with their past values.\n",
    "from mice_inspection_utils import * # Import all functions \n",
    "                                    # NOTICE: this is different from import mice_inspection_utils! In that case you'd have to use the module name as a prefic to access functions, classes or variables. E.g. mice_inspection_utils.some_function()\n",
    "from scipy.stats import linregress\n",
    "from glob import glob # Function that will later be used to find all files that match a pattern\n",
    "\n",
    "# This part avoids possible conflict due to the structure of mice_inspection_utils\n",
    "if not os.path.exists('Data/by_mouse'): \n",
    "    os.makedirs('Data/by_mouse')\n",
    "\n",
    "# GET THE NECESSARY DATA\n",
    "imput_subj_paths = [f\"Data/raw/total.full.rn.RA.total_OTU_table.mouse{i+1}.ovr0.0p.csv\" for i in range(0, 8)]\n",
    "output_subj_paths = [f\"Data/by_mouse/mouse_{i+1}.csv\" for i in range(0, 8)]\n",
    "metadata_path = \"Data/raw/OTU_table.csv\"\n",
    "mi = Mice_Inspection(ip = imput_subj_paths, op = output_subj_paths, mdp = metadata_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Population dynamics of bacteria in the gut microbiome: a data analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Short description\n",
    "\n",
    "The advancement in DNA sequencing tecniques have enabled access to fine-detailed information about the composition of gut microbiota. There is a widespread interest in the scientific comunity around the goal of modeling and predicting the dynamics of these ecological communities, with the aim of producing innovative terapeutic interventions.  However, althogh a wide variety of methods have been published in recent years, many of these are unprincipled and lack reliable validation methods.  Some studies suggest that gut microbiomes cluster into distinct groups (enterotypes), while others argue that these patterns may arise as artifacts of statistical analysis rather than biological reality. \n",
    "\n",
    "\n",
    "In our project, following the article *\"A macroecological description of alternative stable states reproduces intra- and inter-host variability of gut microbiome\", Silvia Zaoli and Jacopo Grilli*, we analyze experimental data of mice gut microbiota assuming a simple stochastic logistic model. This model, specifically, assumes that each bacterial species evolves independently from the others, thus it can be used as a simple, essential null-model to investigate if inter-species interactions can be at all inferred from the data.\n",
    "\n",
    "The logistic model predicts that, after a transient, the species abundances will eventually settle down to a steady state, parametrized by two parametes: the carrying capacity, accounting for the global effect of the environment, and a parameter $\\sigma$ quantifying the stochastic noise. We will follow the methods described in the article, which use a properly defined dissimilarity measure to assess stationarity, designed to overcome the scarsity of data samples.\n",
    "\n",
    "This main task will require preliminary data processing and basic statistic analysis. \n",
    "\n",
    "!!!!   WE HAVEN'T DONE THIS YET !!!!\n",
    "Optionally, in the case where the null model doesn't suffice to explain the data, we will try to infer the interaction matrix of species-species interaction assuming a simple stochastic Lockta Volterra model, and using linear regression with k-fold validation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Preprocessing and sample composition analysis\n",
    "\n",
    "**Data preprocessing**: \n",
    "\n",
    "group OTUs corresponding to same species, fix naming issues, double measures etc, retrieve total number of OTUs, total number of species, set up pandas DataFrames and OTU-species dictionary.\n",
    "\n",
    "**Analysis of the cross-subject sample composition variability at different taxonomic-unit levels:** \n",
    "\n",
    "Biological organisms are classified in groups on the basis of similarity. Biologists refer to these groups as \"taxonomic units\" or taxa. Different taxa are also grouped together in a higher-level group (\"taxon\") based on commonly shared features, resulting in a nested hierarcical classification. The basic scheme of modern classification makes use of $8$ levels of classification, which are, from top to bottom: $\\textit{Domain, Kindom, Phylum, Class, Order, Family, Genus, Species}$. \n",
    "\n",
    "The task is to assess the cross-subject variability of the sample composition at the various levels of the taxonomic classification. The output will be given in the form of stacked-bar plots.\n",
    "\n",
    "**Rank Abundance Distribution (RAD):**\n",
    "    \n",
    "The frequency of each species is computed (= #counts_species/ #counts sample) . Species are ranked with an integer index from the most frequent (i= 1) to the least frequent (i= N). The curve displaying the frequency versus the rank (so-called rank-abundance distribution, or RAD) is experimentally known to assume a lognormal or power-law -like shape, with the first few species occupying the majority of the sample, and a long tail of rare species contributing to the remaining part."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## THE ABOVE PART SHOULD BE REVISITED DEPENDING ON WHAT MIRIAM ACTUALLY WANTS TO TALK ABOUT\n",
    "## OTHERWISE WE STILL NEED TO MODIFY THE ABOVE TEXT BUT WE CAN SIMPLY PRESENT THE DATA ANALISYS AND FUNCTIONS\n",
    "## THAT ARE USED FROM MICE_INSPECT LATER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Theoretical models\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Deterministic Logistic Model\n",
    "\n",
    "The deterministic logistic model represents the growth of a population, or in this case the abundance $\\lambda$ of a microbial species, constrained by the enviroment. \n",
    "The equation for DLM is: $$\\dot{\\lambda} = \\frac{1}{\\tau}\\lambda \\left(1-\\frac{\\lambda}{K}\\right)  $$\n",
    "\n",
    "This model has two parameters:\n",
    "* K: the carrying capacity, it is the maximum population size that an enviroment can sustain, given the available resources.\n",
    "* $\\tau$: ralxation time, it is related to the growth rate of the population and determines the time scale of relaxation to stationarity.\n",
    "\n",
    "1. When $\\lambda$ is small, growth is nearly exponential.\n",
    "2. When $\\lambda$ gets close to K, growth slows.\n",
    "3. When $\\lambda=K$, growth stops as $\\dot{\\lambda}=0$ \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stochastic logistic model\n",
    "\n",
    "The stochastic version of the model introduces some randomness due to enviromental fluctuatons by adding a noise term, which better reflects real-world variability:\n",
    "$$\\dot{\\lambda} = \\frac{1}{\\tau}\\lambda \\left(1-\\frac{\\lambda}{K}\\right)  + \\lambda \\sqrt{\\frac{\\sigma}{\\tau}}\\xi(t) $$\n",
    "where $\\xi(t)$ is the Gaussian white noise.\n",
    "\n",
    "This model has three parameters: \n",
    "* K, $\\tau$ previously described.\n",
    "* $\\sigma$: it measures the intensity of the enviromental noise. \n",
    "\n",
    "The SLM does not include interaction among species and therefore cannot reproduce patterns of interspecies correlation. However, it correctly reproduces several patterns of the dynamics of a single species, including fluctuations and possible extinction events in small populations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lotka-Volterra model --> !!! TO TAKE OFF SINCE NOT DONE\n",
    "\n",
    "The Lotka-volterra model extends population dynamics by incorporating interactions between species, such as competition or mutualism. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Basic time-series analysis\n",
    "\n",
    "Say $\\vec{X}(t) =\\left[x_1(t), x_2(t), \\cdots x_N(t)\\right]$ is the abudance of the first N most abudant species at time t. We model $\\vec{X}(t)$ as a stochastic process, for which\n",
    "each subject $\\vec{X}(t)$ represents a different sample path.\n",
    "\n",
    "* Visualization:\n",
    "\n",
    "    Plot time-series for each mouse and for each species to visualize the evolution in time\n",
    "\n",
    "The basic analysis to perform is:\n",
    "\n",
    "* Compute the time dependent mean and variance $\\mathbb{E}[\\vec{X}(t)],\\, \\mathbb{E}[\\vec{X}^2(t)]$. Plot as a summary the time evolution for the mean +- std of each species\n",
    "* Correlations: filter the most abundant species upon a certain threshold and compute their covariance matrix \n",
    "* Autocorrelation: compute the autocorrelation function $\\mathbb{E}[x(t) \\cdot x(s)]$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ALSO THIS IS MIRIAM'S FIELD ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## JUST AS A COPY OF WHAT THERE IS IN LOGISTICMODEL.IPYNB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species</th>\n",
       "      <th>median_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>1</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>15</th>\n",
       "      <th>17</th>\n",
       "      <th>...</th>\n",
       "      <th>1001</th>\n",
       "      <th>1004</th>\n",
       "      <th>1010</th>\n",
       "      <th>1016</th>\n",
       "      <th>1022</th>\n",
       "      <th>1028</th>\n",
       "      <th>1029</th>\n",
       "      <th>1035</th>\n",
       "      <th>1039</th>\n",
       "      <th>1044</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Prevotella sp. Smarlab 121567</td>\n",
       "      <td>412.0</td>\n",
       "      <td>415.156682</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>220</td>\n",
       "      <td>248</td>\n",
       "      <td>252</td>\n",
       "      <td>286</td>\n",
       "      <td>346</td>\n",
       "      <td>293</td>\n",
       "      <td>277</td>\n",
       "      <td>190</td>\n",
       "      <td>222</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Parabacteroides distasonis</td>\n",
       "      <td>205.5</td>\n",
       "      <td>206.709677</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>145</td>\n",
       "      <td>...</td>\n",
       "      <td>136</td>\n",
       "      <td>157</td>\n",
       "      <td>206</td>\n",
       "      <td>133</td>\n",
       "      <td>177</td>\n",
       "      <td>105</td>\n",
       "      <td>163</td>\n",
       "      <td>147</td>\n",
       "      <td>208</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Barnesiella intestinihominis</td>\n",
       "      <td>142.0</td>\n",
       "      <td>145.419355</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>158</td>\n",
       "      <td>145</td>\n",
       "      <td>184</td>\n",
       "      <td>131</td>\n",
       "      <td>145</td>\n",
       "      <td>154</td>\n",
       "      <td>119</td>\n",
       "      <td>130</td>\n",
       "      <td>71</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Barnesiella viscericola</td>\n",
       "      <td>125.5</td>\n",
       "      <td>137.092166</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "      <td>195</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>85</td>\n",
       "      <td>134</td>\n",
       "      <td>96</td>\n",
       "      <td>126</td>\n",
       "      <td>121</td>\n",
       "      <td>92</td>\n",
       "      <td>71</td>\n",
       "      <td>72</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>Lactobacillus taiwanensis</td>\n",
       "      <td>123.5</td>\n",
       "      <td>174.769585</td>\n",
       "      <td>0</td>\n",
       "      <td>422</td>\n",
       "      <td>309</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>162</td>\n",
       "      <td>282</td>\n",
       "      <td>174</td>\n",
       "      <td>100</td>\n",
       "      <td>306</td>\n",
       "      <td>517</td>\n",
       "      <td>193</td>\n",
       "      <td>513</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 221 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                        species  median_counts  mean_counts  1  \\\n",
       "0           0  Prevotella sp. Smarlab 121567          412.0   415.156682  0   \n",
       "1           1     Parabacteroides distasonis          205.5   206.709677  0   \n",
       "2           3   Barnesiella intestinihominis          142.0   145.419355  0   \n",
       "3           4        Barnesiella viscericola          125.5   137.092166  0   \n",
       "4           2      Lactobacillus taiwanensis          123.5   174.769585  0   \n",
       "\n",
       "     8    9  10   15   17  ...  1001  1004  1010  1016  1022  1028  1029  \\\n",
       "0    0    0   2   10   18  ...   220   248   252   286   346   293   277   \n",
       "1    0    0   2   41  145  ...   136   157   206   133   177   105   163   \n",
       "2    0    0   1   19   58  ...   158   145   184   131   145   154   119   \n",
       "3    1    0   1  112  195  ...    66    85   134    96   126   121    92   \n",
       "4  422  309   0   48   15  ...    36   162   282   174   100   306   517   \n",
       "\n",
       "   1035  1039  1044  \n",
       "0   190   222   364  \n",
       "1   147   208   222  \n",
       "2   130    71   161  \n",
       "3    71    72   147  \n",
       "4   193   513   175  \n",
       "\n",
       "[5 rows x 221 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# By calling mice_df we have the dataframes for each mouse\n",
    "df = mi.mice_df[0]\n",
    "df.head() # -> As we can see we have the bacterial species, the median, the mean counts, and the counts per day (at day 1, 8, 9 ...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>day</th>\n",
       "      <th>mouse_1</th>\n",
       "      <th>mouse_2</th>\n",
       "      <th>mouse_3</th>\n",
       "      <th>mouse_4</th>\n",
       "      <th>mouse_5</th>\n",
       "      <th>mouse_6</th>\n",
       "      <th>mouse_7</th>\n",
       "      <th>mouse_8</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.744024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.353553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.755929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.707107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.875</td>\n",
       "      <td>2.474874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.750</td>\n",
       "      <td>20.940392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   day  mouse_1  mouse_2  mouse_3  mouse_4  mouse_5  mouse_6  mouse_7  \\\n",
       "0    0        0        1        0        0        0        0        0   \n",
       "1    1        0        0        0        0        0        0        0   \n",
       "2    2        0        1        1        0        0        0        0   \n",
       "3    3        0        0        0        0        2        0        0   \n",
       "4    4        0        0        0        0        0        0        0   \n",
       "5    5        0        0        0        0        0        0        0   \n",
       "6    6        0        0        0        0        0        0        0   \n",
       "7    7        0        0        0        0        0        0        0   \n",
       "8    8        0        0        0        0        0        0        0   \n",
       "9    9        0        0        9        0        1       60        0   \n",
       "\n",
       "   mouse_8   mean        std  \n",
       "0        2  0.375   0.744024  \n",
       "1        1  0.125   0.353553  \n",
       "2        2  0.500   0.755929  \n",
       "3        0  0.250   0.707107  \n",
       "4        0  0.000   0.000000  \n",
       "5        0  0.000   0.000000  \n",
       "6        0  0.000   0.000000  \n",
       "7        0  0.000   0.000000  \n",
       "8        7  0.875   2.474874  \n",
       "9        0  8.750  20.940392  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# By calling get_species_df\n",
    "df = mi.get_species_df(species = 'Prevotella sp. Smarlab 121567')\n",
    "df.head(10) # -> As we can see we have the count per day of the selected species in each mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Stochastic Logistic Model\n",
    "\n",
    "Here we follow the main reference *\"A macroecological description of alternative stable states reproduces intra- and inter-host variability of gut microbiome\", Silvia Zaoli and Jacopo Grilli*."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dissimilarity\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
